File: 11_phonetic_algorithms/src/soundex.py
"""
Soundex algorithm implementation.
Based on: https://code.activestate.com/recipes/52213-soundex-algorithm
Adapted to match: https://www.archives.gov/research/census/soundex
"""

# soundex values for the alphabet
_DIGITS = "0123012H02245501262301W202"
_LENGTH = 4


def soundex(
    input: str,
    digits: str = _DIGITS,
    length: int = _LENGTH,
) -> str:
    output = ""
    char_0 = ""

    # translate alpha characters in input to soundex digits
    for char in input.upper():
        if char.isalpha():
            # remember first letter
            if not char_0:
                char_0 = char
            digit = digits[ord(char) - ord("A")]
            # duplicate consecutive soundex digits are skipped
            if not output or (digit not in [output[-1], "H", "W"]):
                output += digit

    # replace first digit with first alpha character
    output = char_0 + output[1:]

    # remove all 0s from the soundex code
    output = output.replace("0", "")

    # return soundex code padded to length characters
    return (output + length * "0")[:length]

File: 11_phonetic_algorithms/src/soundex.dart
/// Soundex algorithm implementation (in Dart).
/// Based on: https://code.activestate.com/recipes/52213-soundex-algorithm
/// Adapted to match: https://www.archives.gov/research/census/soundex

// soundex values for the alphabet
const _DIGITS = "0123012H02245501262301W202";
const _LENGTH = 4;

const _CHAR_0 = 48;
const _CHAR_A = 65;
const _CHAR_Z = 90;

String soundex(String input, {String digits = _DIGITS, int length = _LENGTH}) {
  var output = "";
  var firstLetter = "";

  for (final char in input.toUpperCase().runes) {
    if (char < _CHAR_A || char > _CHAR_Z) {
      continue;
    }

    if (firstLetter.isEmpty) {
      firstLetter = String.fromCharCode(char);
      output += firstLetter;
      continue;
    }

    final digit = digits.codeUnitAt(char - _CHAR_A);
    if (digit != _CHAR_0 && digit != output.codeUnitAt(output.length - 1)) {
      output += String.fromCharCode(digit);
    }
    
    if (output.length == length) {
      break;
    }
  }

  return output.padRight(length, '0');
}
